lib/remap-lang/src/expression/assignment.rs

```diff
@@ -0,0 +1,50 @@
+use super::Error as E;
+use crate::{Expr, Expression, Object, Result, State, Value};
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum Error {
+    #[error("unable to insert value in path: {0}")]
+    PathInsertion(String),
+}
+
+#[derive(Debug)]
+pub(crate) enum Target {
+    Path(String),
+    Variable(String),
+}
+
+#[derive(Debug)]
+pub(crate) struct Assignment {
+    target: Target,
+    value: Box<Expr>,
+}
+
+impl Assignment {
+    pub fn new(target: Target, value: Box<Expr>) -> Self {
+        Self { target, value }
+    }
+}
+
+impl Expression for Assignment {
+    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>> {
+        let value = self.value.execute(state, object)?;
+
+        match value {
+            None => Ok(None),
+            Some(value) => {
+                match &self.target {
+                    Target::Variable(ident) => {
+                        state
+                            .variables_mut()
+                            .insert(ident.to_owned(), value.clone());
```

Comment:
I recognize the value of having expressions usable as a value, but these clones concern me for run-time efficiency, considering the value may be a deep structure (think `Value::Map` which may contain further maps, etc). I'm not sure off-hand how to handle it better though, and I recognize such a change would affect many return types. We'd like to return `Result<Option<&Value>>` but lifetimes almost certainly prevent that. Possibly something with `Rc<Value>`?

Reply:
I imagine most of the sizable data will ultimately be `bytes::Bytes`, which is very cheap to clone. It might not be as expensive as we think?

Reply:
True. The most *frequently* cloned-after-assignment data would be scalars, all of which are quick to clone (thanks to `Bytes` for strings). However, we also want to move parsing functionality like `parse_json` into remap, and that will not be an inexpensive clone.

Reply:
Yeah, I considered returning `&Value`, but decided against that change for now, as — as you mention — that has quite some impact in the implementation, and potentially needs reference counting.

I'm going to run the existing benchmarks (and potentially expand them) once this is integrated back into Vector, and there's room for improvement here, but I think we're fine for now with this clone.

_(as an aside, I removed variable support for now, so cloning doesn't happen here, but it does still happen in other places, so the point still stands)_